src/libcore/borrow.rs

```diff
@@ -12,26 +12,143 @@
 
 #![stable(feature = "rust1", since = "1.0.0")]
 
-/// A trait for borrowing data.
+/// A trait identifying how borrowed data behaves.
 ///
-/// In general, there may be several ways to "borrow" a piece of data.  The
-/// typical ways of borrowing a type `T` are `&T` (a shared borrow) and `&mut T`
-/// (a mutable borrow). But types like `Vec<T>` provide additional kinds of
-/// borrows: the borrowed slices `&[T]` and `&mut [T]`.
+/// If a type implements this trait, it signals that a reference to it behaves
+/// exactly like a reference to `Borrowed`. As a consequence, if a trait is
+/// implemented both by `Self` and `Borrowed`, all trait methods that
+/// take a `&self` argument must produce the same result in both
+/// implementations.
```

Comment:
This is too strong a statement in my opinion. For example the following is probably fine. Is there a way to make a stronger statement than the existing documentation but without applying it to "all trait methods"?

```rust
#![feature(get_type_id)]

use std::any::Any;
use std::borrow::Borrow;

fn assert_borrow<Q: ?Sized, K: Borrow<Q>>() {}

fn main() {
    assert_borrow::<str, String>();
    let str_id = <str as Any>::get_type_id("");
    let string_id = <String as Any>::get_type_id(&String::new());
    println!("{}", str_id == string_id); // false
}
```

Reply:
This exception could perhaps be formulated as ‘trait methods that concern the type of a value rather than a value.’ It’s a bit awkward; perhaps there is a better way to formulate this?

Are there any exceptions where the value itself concerned, though?

Reply:
Here is one that proclaims to involve the "val" of a value.

```rust
use std::mem;

trait Layout {
    fn size_of_val(&self) -> usize {
        mem::size_of_val(self)
    }
    fn align_of_val(&self) -> usize {
        mem::align_of_val(self)
    }
}
impl<T: ?Sized> Layout for T {}

fn main() {
    println!("size of str: {}", <str as Layout>::size_of_val(""));
    println!("align of str: {}", <str as Layout>::align_of_val(""));
    println!("size of String: {}", <String as Layout>::size_of_val(&String::new()));
    println!("align of String: {}", <String as Layout>::align_of_val(&String::new()));
}
```

Reply:
Here is one from the standard library that behaves differently between Q and K and the value is concerned.

```rust
use std::borrow::Borrow;

fn assert_borrow<Q: ?Sized, K: Borrow<Q>>() {}

fn main() {
    assert_borrow::<str, &str>();

    // copies the data
    let _: String = <str as ToOwned>::to_owned("...");

    // does not copy the data
    let _: &str = <&str as ToOwned>::to_owned(&"...");
}
```

Reply:
Perhaps distinguish between ‘management of data’ and ‘properties of data’? It is too vague for my liking, but it sounds like there won’t be a hard-fast rule and it will always be a judgment call.